File: trees/BIT.py
# @ref https://oi.wiki/ds/fenwick

n=...
c=[0]*(n+1)

lowbit=lambda x: x&-x

def prep(a):
    res=[0]*(n+1)
    s=0
    for i in range(1,n+1):
        s+=a[i]
        res[i]=s
    return res

def init(a):
    pa=prep(a)
    for i in range(1,n+1):
        c[i]=pa[i]-pa[i-lowbit(i)]

def getsum(x):
    ans=0
    while x:
        ans+=c[x]
        x-=lowbit(x)
    return ans

def add(x,k):
    while x<=n:
        c[x]+=k
        x+=lowbit(x)

File: trees/BIT.cc
template<typename T>
struct BIT {
	int n;
	vector<T> c;
	BIT(size_t n) : n(n), c(n + 1) {}
	void add(size_t i, const T& k) {
		while (i <= n) {
			c[i] += k;
			i += lowbit(i);
		}
	}
	T getsum(size_t i) {
		T res = {};
		while (i) {
			res += c[i];
			i -= lowbit(i);
		}
		return res;
	}
};
